flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -766,13 +776,12 @@ private void returnExcessMemorySegments() {
 
     @GuardedBy("availableMemorySegments")
     private boolean hasExcessBuffers() {
-        return numberOfRequestedOverdraftMemorySegments > 0
-                || numberOfRequestedMemorySegments > currentPoolSize;
+        return numberOfRequestedOverdraftMemorySegments > 0;
     }
 
     @GuardedBy("availableMemorySegments")
     private boolean isRequestedSizeReached() {
-        return numberOfRequestedMemorySegments >= currentPoolSize;
+        return numberOfRequestedMemorySegments == currentPoolSize;
```

Comment:
I think you should not change this condition. Since, in my opinion, the rule of thumb for concurrent code looks something like this: "If the value physically can be greater than the border value we should also compare it with the border as >= even if logically it can not be greater". 
In this case, we have the moment of time when `numberOfRequestedMemorySegments > currentPoolSize` it is under the synchronization but anyway.

Reply:
> In my opinion, the rule of thumb for concurrent code looks something like this: "If the value physically can be greater than the border value we should also compare it with the border as >= even if logically it can not be greater".

This rule really make sense to me and thanks for telling me about this! Maybe I was a little too radical before. After all, the world full of concurrency is not safe at all. 😞